pkg/server/server_test.go

```diff
@@ -518,6 +518,77 @@ func TestThreeProngThroughVariousLayers(t *testing.T) {
 	}
 }
 
+func TestCheckObjectsThrottledTimeout(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+
+	_, ds, _ := util.MustBootstrapDatastore(t, "memory")
+	s := MustNewServerWithOpts(
+		WithDatastore(ds),
+		WithDispatchThrottlingCheckResolverFrequency(5*time.Millisecond),
+		WithDispatchThrottlingCheckResolverEnabled(true),
+		WithDispatchThrottlingCheckResolverThreshold(5),
+	)
+	t.Cleanup(func() {
+		s.Close()
+	})
+
+	createStoreResp, err := s.CreateStore(context.Background(), &openfgav1.CreateStoreRequest{
+		Name: "openfga-test",
+	})
+	require.NoError(t, err)
+
+	storeID := createStoreResp.GetId()
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+	schema 1.1
+
+type user
+
+type group
+  relations
+    define member: [user, group#member]
+`)
+
+	writeAuthModelResp, err := s.WriteAuthorizationModel(context.Background(), &openfgav1.WriteAuthorizationModelRequest{
+		StoreId:         storeID,
+		SchemaVersion:   model.GetSchemaVersion(),
+		TypeDefinitions: model.GetTypeDefinitions(),
+	})
+	require.NoError(t, err)
+
+	modelID := writeAuthModelResp.GetAuthorizationModelId()
+
+	_, err = s.Write(context.Background(), &openfgav1.WriteRequest{
+		StoreId: storeID,
+		Writes: &openfgav1.WriteRequestWrites{
+			TupleKeys: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:x", "member", "group:1#member"),
+				tuple.NewTupleKey("group:x", "member", "group:2#member"),
+				tuple.NewTupleKey("group:x", "member", "group:3#member"),
+				tuple.NewTupleKey("group:x", "member", "group:4#member"),
+				tuple.NewTupleKey("group:x", "member", "group:5#member"),
+				tuple.NewTupleKey("group:x", "member", "group:6#member"),
+				tuple.NewTupleKey("group:x", "member", "group:7#member"),
+				tuple.NewTupleKey("group:x", "member", "group:8#member"),
+				tuple.NewTupleKey("group:x", "member", "group:9#member"),
+				tuple.NewTupleKey("group:x", "member", "group:10#member"),
+			},
+		},
+	})
+	require.NoError(t, err)
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Millisecond)
```

Comment:
Seems this test has been written largely based on luck - the assumption is that the 10 dispatches for the 10 different groups causes > 20ms delay.

I may advise that we just mock the storage layer and eliminate the memory datastore in this case so it's more of a controlled test.

We should set dispatch throttling threshold to 1, and mock the storage read(s) with a bit of latency, and then assert the throttled error. This way it is more deterministic and not so co-dependent on time.

Reply:
This actually is not based on luck.  
We know that this will take 10 dispatches will means that there are 5 dispatch being throttled.  Each throttling will take at least 5 milliseconds.  Therefore, the whole request must take at least 25 milliseconds to complete (5 * 5ms = 25ms).  Hence, this is always less than 20ms and must have timeout.

Perhaps I will update some comments to make it very clear.

Reply:
But that's still based on the tick frequency which we know doesn't necessarily tick super accurately below 10ms range. So the assumption that 5ms overhead is plenty is a bit of an assumption. That's what I mean by "luck".

But it's ok, no need to sweat it. If we see flakiness we can adjust accordingly.